widget/widget_date_input.go

```diff
@@ -0,0 +1,208 @@
+	t.yearSeg.Text = t.currentDate.Format("2006")
+
+	t.daySeg.Style.TextStyle.Bold = false
+	t.monthSeg.Style.TextStyle.Bold = false
+	t.yearSeg.Style.TextStyle.Bold = false
+
+	if t.selectSection == 1 {
+		if t.curSection == "d" {
+			t.daySeg.Style.TextStyle.Bold = true
+		} else if t.curSection == "m" {
+			t.monthSeg.Style.TextStyle.Bold = true
+		} else if t.curSection == "y" {
+			t.yearSeg.Style.TextStyle.Bold = true
+		}
+	}
+	t.richText.Refresh()
+}
+
+// Start: implement focusable interface
+func (t *JDateInputWidget) FocusGained() {
+	t.border.FillColor = theme.FocusColor()
+	t.border.StrokeColor = theme.PrimaryColor()
+	t.updateDisplay()
+	t.Refresh()
+}
+
+func (t *JDateInputWidget) FocusLost() {
+	t.border.FillColor = theme.BackgroundColor()
+	t.border.StrokeColor = theme.InputBorderColor()
+	t.daySeg.Style.TextStyle.Bold = false
+	t.monthSeg.Style.TextStyle.Bold = false
+	t.yearSeg.Style.TextStyle.Bold = false
+	t.Refresh()
+}
+
+func (t *JDateInputWidget) TypedRune(k rune) {
+	// needed
+}
+
+func (t *JDateInputWidget) TypedKey(e *fyne.KeyEvent) {
+	if e.Name == fyne.KeyUp || e.Name == fyne.KeyU {
+		t.addTime(t.curSection, 1)
+		t.updateDisplay()
+	} else if e.Name == fyne.KeyDown || e.Name == fyne.KeyD {
+		t.addTime(t.curSection, -1)
+		t.updateDisplay()
+	} else if e.Name == fyne.KeyRight || e.Name == fyne.KeyR {
+		if t.curSection == "d" {
+			t.curSection = "m"
+		} else if t.curSection == "m" {
+			t.curSection = "y"
+		} else if t.curSection == "y" {
+			t.curSection = "d"
+		}
+		t.updateDisplay()
+	} else if e.Name == fyne.KeyLeft || e.Name == fyne.KeyL {
+		if t.curSection == "d" {
+			t.curSection = "y"
+		} else if t.curSection == "m" {
+			t.curSection = "d"
+		} else if t.curSection == "y" {
+			t.curSection = "m"
+		}
+		t.updateDisplay()
+	} else if e.Name == fyne.KeySpace {
+		t.currentDate = time.Now()
+		t.updateDisplay()
+	} else if e.Name == fyne.KeyPageDown {
+		t.addTime("d", -10)
+		t.updateDisplay()
+	} else if e.Name == fyne.KeyPageUp {
+		t.addTime("d", 10)
+		t.updateDisplay()
+	}
+}
+
+// End: implement focusable interface
+
+// Start: implemnet Tappable
+func (t *JDateInputWidget) Tapped(p *fyne.PointEvent) {
+	t.mainWindow.Canvas().Focus(t)
+}
+
+// End: implemnet Tappable
+
+// helper functions
+func newTextSegment() *widget.TextSegment {
+	t := new(widget.TextSegment)
+	t.Style.Inline = true
+	return t
+}
+
+// public interface
+func (t *JDateInputWidget) SetDate(newDate time.Time) {
+	t.currentDate = newDate
+	t.updateDisplay()
+}
+
+func (t *JDateInputWidget) GetDate() time.Time {
+	return t.currentDate
+}
+
+func (t *JDateInputWidget) GetDateString() string {
+	return t.currentDate.Format("02-Jan-2006")
+}
+
+// widget render
+func (t *JDateInputWidget) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	return widget.NewSimpleRenderer(t.gl)
+}
+
+// constructor
+func NewJDateInputWidget(app fyne.App, window fyne.Window) *JDateInputWidget {
```

Comment:
All exported functions and methods should have a comment describing if them. See already existing code in this repo for examples.

Reply:
The comments also need to follow the Go standard starting with the name of the function.